### PR TITLE
MGMT-13733: Add `generate-upgrade-bundle` command

### PR DESCRIPTION
**Pull request by jhernand**
This patch adds a new `generate-upgrade-bundle` command that generates a tarball containing all the images that are necessary to perform an upgrade without a registry.

Related: https://issues.redhat.com/browse/MGMT-13733

### CODE REVIEW

**Comment by jhernand** [comment]
@danielerez @nmagnezi this is the first step to move the code from the [prototype](https://github.com/jhernand/upgrade-tool) to this repository. It moves the `upgrade-tool create bundle` command to `openshift-appliance generate-upgrade-bundle`.

The next step is to move also the controller and associated tools, see #89.

File: pkg/asset/upgrade/bundle.go [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+		Build()
+	if err != nil {
+		return err
+	}
+
+	// Create the temporary directory for the bundle contents:
+	tmpDir := filepath.Join(envCfg.TempDir, "upgrade")
+	err = os.Mkdir(tmpDir, 0700)
+	if errors.Is(err, os.ErrExist) {
+		err = nil
```
Comment (danielerez) [inline pkg/asset/upgrade/bundle.go:82]: We can ask the user to run 'clean' command first in the case, as done in build: https://github.com/openshift/appliance/blob/ae232a33d1c23b938dbff0b15ced32b9e88525d2/cmd/build.go#L45

Just need to add 'upgrade' dir to the clean command (so worth extracting to a constant).

File: pkg/asset/upgrade/bundle.go [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+	jqTool, err := jq.NewQueryTool().
+		SetLogger(logrusr.New(logger)).
+		Build()
+	if err != nil {
+		return err
+	}
+
+	// Create the temporary directory for the bundle contents:
+	tmpDir := filepath.Join(envCfg.TempDir, "upgrade")
+	err = os.Mkdir(tmpDir, 0700)
```
Comment (danielerez) [inline pkg/asset/upgrade/bundle.go:80]: Can we use os.ModePerm?

File: pkg/asset/upgrade/bundle.go [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+// Generate generates the upgrade bundle.
+func (b *Bundle) Generate(dependencies asset.Parents) error {
+	// Get the logger:
+	logger := logrus.StandardLogger()
+
+	// Get the dependencies:
+	envCfg := &config.EnvConfig{}
+	appCfg := &config.ApplianceConfig{}
+	dependencies.Get(envCfg, appCfg)
+
```
Comment (danielerez) [inline pkg/asset/upgrade/bundle.go:58]: Is there anything worth saving in cache dir? If so, need to first check if already exists using envConfig.FindInCache.

File: pkg/jq/query_tool.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+// QueryString is similar to Query, but it expects an input string containing JSON text.
+func (t *QueryTool) QueryString(query string, input string, output any) error {
+	return t.QueryBytes(query, []byte(input), output)
+}
+
+// QueryBytes is similar to Query, but it expects as input an array of bytes containing the JSON
+// text.
+func (t *QueryTool) QueryBytes(query string, input []byte, output any) error {
```
Comment (danielerez) [inline pkg/jq/query_tool.go:69]: I just wonder if there's already a library the does all this work?:/ Need to search..

File: pkg/asset/upgrade/bundle.go [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+	// Use the 'oc adm release info' command to extract the release metadata JSON from the
+	// release metadata image:
+	ocPath, err := exec.LookPath("oc")
+	if err != nil {
+		return
+	}
+	ocStdout := &bytes.Buffer{}
+	ocStderr := &bytes.Buffer{}
+	ocCmd := exec.Cmd{
```
Comment (danielerez) [inline pkg/asset/upgrade/bundle.go:218]: Can you use GetImageFromRelease here? See pkg/release/release.go

File: pkg/asset/upgrade/bundle.go [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+	)
+
+	// Extract from the release metadata JSON the references of the rest of the images that are
+	// part of the release:
+	type ImageInfo struct {
+		Tag string `json:"tag"`
+		Ref string `json:"ref"`
+	}
+	var imageInfos []ImageInfo
+	err = g.jqTool.QueryBytes(
```
Comment (danielerez) [inline pkg/asset/upgrade/bundle.go:264]: Can we maybe use mapping.txt file that oc-mirror is creating: https://github.com/openshift-kni/telco-ran-tools/blob/main/cmd/download.go#L345C1-L345C1
Might save some work here...

File: pkg/asset/upgrade/bundle.go [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+	err = os.WriteFile(authFile, []byte(g.appCfg.Config.PullSecret), 0400)
+	if err != nil {
+		return err
+	}
+
+	// Download the images:
+	imageTags := maps.Keys(imageMap)
+	slices.Sort(imageTags)
+	registryAddr := registryServer.ListenAddr()
+	for i, srcTag := range imageTags {
```
Comment (danielerez) [inline pkg/asset/upgrade/bundle.go:331]: just to make sure I'm following. iiuc, the work here is something like this: https://github.com/openshift-kni/telco-ran-tools/blob/main/cmd/download.go#L482

File: pkg/asset/upgrade/bundle.go [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+				"file": bundleFile,
+			}).Error("Failed to close bundle file")
+		}
+	}()
+	_, err = io.Copy(bundleHash, bundleReader)
+	if err != nil {
+		return err
+	}
+	bundleSum := hex.EncodeToString(bundleHash.Sum(nil))
+	digestWriter, err := os.OpenFile(digestFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
```
Comment (danielerez) [inline pkg/asset/upgrade/bundle.go:451]: Why do we need to output a digest file? I.e. where should it be used later on?

**Review by danielerez — COMMENTED** [review]
Partial review. It's a Long PR:)

**Pull request closed without merging**